wallet-core/src/transaction.rs

```diff
@@ -0,0 +1,336 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+//! Implementations of basic wallet functionalities to create transactions.
+
+use alloc::vec::Vec;
+
+use rand::{CryptoRng, RngCore};
+
+use ff::Field;
+use poseidon_merkle::Opening;
+use zeroize::Zeroize;
+
+use execution_core::{
+    signatures::bls::SecretKey as BlsSecretKey,
+    stake::{Stake, Withdraw as StakeWithdraw, STAKE_CONTRACT},
+    transfer::{
+        contract_exec::{ContractCall, ContractExec},
+        phoenix::{
+            Note, Prove, PublicKey as PhoenixPublicKey,
+            SecretKey as PhoenixSecretKey, Transaction as PhoenixTransaction,
+            NOTES_TREE_DEPTH,
+        },
+        withdraw::{Withdraw, WithdrawReceiver, WithdrawReplayToken},
+        Transaction,
+    },
+    BlsScalar, ContractId, JubJubScalar,
+};
+
+/// Types that are client of the prover.
+pub trait ProverClient {
+    /// Error returned by the node client.
+    type Error;
+
+    /// Requests that a node prove the given unproven [`PhoenixTransaction`] and
+    /// later propagates it.
+    fn compute_proof_and_propagate(
+        utx: &PhoenixTransaction,
+    ) -> Result<Transaction, Self::Error>;
+}
+
+/// Create a [`Transaction`] that is paid in phoenix-notes.
+#[allow(clippy::too_many_arguments)]
+pub fn phoenix_transaction<R: RngCore + CryptoRng, PC: ProverClient>(
+    rng: &mut R,
+    sender_sk: &PhoenixSecretKey,
```

Comment:
a type alias here would be nice, its typed out multiple times too

Reply:
type-alias for what?

Reply:
```rust

type Input = (Note, Opening<_', _'>);`
type Inputs = Vec<Input>;

```

Reply:
Hm, I'm not sure if creating a type for that is worth it as sometimes a `Vec<Note, Opening<..>>` is needed but sometimes we also have a `Vec<Note, BlsScalar>` (for the nullifier) or even a `Vec<Note, Opening<..>, BlsScalar>`...

That said, I totally agree that this reads horribly.
I think the best thing to do is to re-export the `poseidon-merkle` types through `execution-core`. This which will already make this read much better (something like `Vec<Note, NoteOpening>`, `Vec<Note, NoteOpening, BlsScalar>`).